lib/statsd/instrument/strict.rb

```diff
@@ -111,31 +107,113 @@ def collect_metric(type, name, value, sample_rate:, tags: nil, prefix:, metadata
     end
 
     module StrictMetaprogramming
-      def statsd_measure(method, name, sample_rate: nil, tags: nil,
-        prefix: StatsD.prefix, no_prefix: false, as_dist: false)
-
+      def statsd_measure(method, name, sample_rate: nil, tags: nil, no_prefix: false)
         check_method_and_metric_name(method, name)
-        super
+
+        # Unfortunately, we have to inline the new method implementation ebcause we have to fix the
+        # Stats.measure call to not use the `as_dist` and `prefix` arguments.
```

Comment:
I think you mean

```suggestion
        # statsd_measure call to not use the `as_dist` and `prefix` arguments.
```

Although, couldn't we change the non-strict version of `statsd_measure` to use a keyword argument splat to pass through deprecated parameters?  If it is passing those parameters to a strict singleton method, then I don't think we would even have to override these method wrappers for strictness, unless we wanted to make sure that the arguments are checked at initialization time.

Reply:
For example, we could do this to avoid passing `prefix` or `as_dist` arguments while still handling deprecated positional arguments in the same way.

```diff
diff --git a/lib/statsd/instrument.rb b/lib/statsd/instrument.rb
index d32b34d..14412eb 100644
--- a/lib/statsd/instrument.rb
+++ b/lib/statsd/instrument.rb
@@ -90,15 +90,14 @@ module StatsD
     #    callable to dynamically generate a metric name
     # @param metric_options (see StatsD#measure)
     # @return [void]
-    def statsd_measure(method, name, deprecated_sample_rate_arg = nil, deprecated_tags_arg = nil, as_dist: false,
-      sample_rate: deprecated_sample_rate_arg, tags: deprecated_tags_arg, prefix: nil, no_prefix: false)
+    def statsd_measure(method, name, deprecated_sample_rate_arg = nil, deprecated_tags_arg = nil,
+      sample_rate: deprecated_sample_rate_arg, tags: deprecated_tags_arg, **statsd_kwargs)
 
       add_to_method(method, name, :measure) do
         define_method(method) do |*args, &block|
           key = StatsD::Instrument.generate_metric_name(name, self, *args)
-          prefix ||= StatsD.prefix
           StatsD.measure(
-            key, sample_rate: sample_rate, tags: tags, prefix: prefix, no_prefix: no_prefix, as_dist: as_dist
+            key, sample_rate: sample_rate, tags: tags, **statsd_kwargs
           ) do
             super(*args, &block)
           end
```

Reply:
It would be a minor performance hit right, because we would use a keyword splat? It does clean up the code quite nicely.

Reply:
Oh, it looks like you are right that it would be a minor performance hit.   I expected that capturing the keyword arguments would allocate a hash, but that happens at initialization time.  But it looks like applying the keyword arguments hash actually results in a couple of allocations.

That seems like something that could be optimized in ruby itself.  It is unfortunate that we have a trade-off between code and runtime complexity with the current state of ruby.

Reply:
Yup, for the time being this strict mode file is pretty awful, but at least it's contained to this file. I'd prefer the 💩 not to spread outside of this file, even if that means to take on some extra duplication.